docs/markdown/Builtin-options.md

```diff
@@ -206,14 +206,22 @@ available on all platforms or with all compilers:
 | b_pch               | true                 | true, false                                                   | Use precompiled headers                                                        |
 | b_pgo               | off                  | off, generate, use                                            | Use profile guided optimization                                                |
 | b_sanitize          | none                 | see below                                                     | Code sanitizer to use                                                          |
+| b_sanitizers        | []                   | see below                                                     | Code sanitizer to use                                                          |
 | b_staticpic         | true                 | true, false                                                   | Build static libraries as position independent                                 |
 | b_pie               | false                | true, false                                                   | Build position-independent executables (since 0.49.0)                          |
 | b_vscrt             | from_buildtype       | none, md, mdd, mt, mtd, from_buildtype, static_from_buildtype | VS runtime library to use (since 0.48.0) (static_from_buildtype since 0.56.0)  |
 
-The value of `b_sanitize` can be one of: `none`, `address`, `thread`,
-`undefined`, `memory`, `leak`, `address,undefined`, but note that some
-compilers might not support all of them. For example Visual Studio
-only supports the address sanitizer.
+† The default and possible values of sanitizers changed in 1.4. Before 1.4 they
+were string values, and restricted to a specific subset of values: `none`,
+`address`, `thread`, `undefined`, `memory`, `leak`, or `address,undefined`. In
+1.4 it was changed to a free form array of sanitizers, which are checked by a
+compiler and linker check. For backwards compatibility reasons
+`get_option('b_sanitize')` continues to return a string value, which is
+guaranteed to match the old string value if the array value contains the same
+values (ie, `['undefined', 'address'] == 'address,undefined`). I a value not
+allowed before 1.4 is used `b_sanitize` will return a string in undefined order.
+A new value of `b_sanitizers` can be used for >= 1.4 to get a free form array,
+with no ordering guarantees.
```

Comment:
Could we do something like if the meson_version defined by the project is >= 1.4.0, return an array, else return string?

Reply:
That's what I did initially and others were absolutely against that. Having done CMake, I can see the concern

Reply:
Any reason you didn't want to continue with the b_sanitizers option? I think I would rather just have a new option than an array option that returns a string.

Reply:
I would rather not add extra complexity of having mutually exclusive options if we don't need them, in this case b_sanitize already accepts the same format that b_sanitizers would return anyway. If you don't like the `get_option('b_sanitizers')`, perhaps we could come up with something different there? like a new keyword for `get_option()` like, `get_option('b_sanitize', version : 2)` or something like that?

Reply:
What is the CMake concern that you have in mind?

I don't think the version number is a good idea, it adds a new layer of opt-ins that is unnecessary when we already use the project's Meson version for deprecations.

I think b_sanitizers is a better name than b_sanitize, so a rename is fine if you want to kick the can a little bit further down the road (in the sense that I cannot think of any such new name for `b_pie`). However, the rename should be done contextually with deprecation of `b_sanitize`; `-Db_sanitize` should be automatically converted to `-Db_sanitizers`, with the old name surviving only as an argument to `get_option()`.

With respect to the order, I would return the sanitizers in sorted order for `get_option('b_sanitize')`, since it's a simple generalization of the special case `address,undefined`.

Reply:
> What is the CMake concern that you have in mind?

@eli-schwartz might have more on this. For me, the concern is that CMake policies (which is essentially what's being proposed via minimum version checks), change the behavior of existing functions such that you don't inherently know what they're going to do. Using an explicit keyword argument makes you expect a different result.

> I think b_sanitizers is a better name than b_sanitize, so a rename is fine if you want to kick the can a little bit further down the road (in the sense that I cannot think of any such new name for b_pie).

We have never renamed things in the past because of a change from a singular value to an array. for example `add_project_arguments(language : _)` originally only accepted a single string for one language. It was later changed to accept a list, but was left as `language`. `b_sanitizers` is a better name, but I don't think it's worth the pain from an implementation and user point of view of having two different options that are mutually exclusive, map over each other, and then have to track two diferent `get_option()` values. It's a lot of pain to add and `s` to the end of a word, when we have never done that in the past.

> With respect to the order, I would return the sanitizers in sorted order for get_option('b_sanitize'), since it's a simple generalization of the special case `address,undefined`

That's what I've done, but I don't want to require that, so I'm only defined that `address,undefined` will be guaranteed, otherwise its an implementation detail. Muon might decide sorting all the time isn't as easy as it is for Python, and want to special case it differently.

Reply:
The problem with CMake policies lies with how they interact with third party functions that are liberally included in the sources. Meson doesn't have them, and for the first time I see the point in not having them. Instead, the "policy" can be specified per subproject, which is Meson's prescribed unit for embedding external code. This means that the project's minimum meson version does not have the issues that CMake policies have.

If you don't want that, perhaps a "type" keyword argument to get_option could be added:

* all options support the option type. It is the default for all options except b_sanitize and (later) b_pie
* all options support string. For b_sanitize only, the order is unspecified other than the `address,undefined` case, in order to address backwards compatibility
* in addition to its actual type "feature", b_pie supports boolean and has it as the default.

However, I still believe that this is overengineered, compared to a heavily simplified and restricted "policy".

Reply:
I guess I can go either for a version based policy or for the keyword argument based policy (let's paint the bikeshed if we agree to go that route). I really dislike the need to create new names if the types of builtin options changes over time, and would prefer one of the policy based ideas.

@eli-schwartz you've been vocal about this in the past, do you have a strong opinion on which way?

Reply:
Ping... would like to have a resolution here and for #12617.

I really dislike the `version` argument. I think Meson so far has successfully used a single (more accurately, per subproject) version number, and I see no reason to change this.

In fact, considering that `get_option('b_sanitize')` or `get_option('b_pie')` would likely be accessed only once in meson.build, adding (and requiring) a "version" argument there seems closer to CMake fine-grained policies, than just relying on the current subproject's minimum version requirement.

Reply:
The problem with a minimum version requirement is that up until now, the minimum version requirement was guaranteed to have no effect other than warnings control.

And not all projects use them at all! Which is "fine", but users get no real visibility into what version of meson you need, so they'd better upgrade regardless of which version of meson is provided by `apt install`.

Changing runtime behavior based on this, means that:
- people who bump their minimim version proactively may suddenly have a crash, and worse, that crash may be in conditional code that they didn't hit in local testing
- people who don't specify a minimum version either crash or don't crash, depending on which version of meson you use. CMake policies "don't have that issue" because the policy is separate from version and you opt into each one you want, and if you don't then you're just out of luck because you were expected to run a debug execution of cmake that reports deprecation warnings, then update your code (but not third-party functions) and flip the toggle and pray no third-party functions break

In contrast, renaming the option is the conservatively safe approach and *cannot* break projects, no matter how unlikely. And leaving aside developer implementation challenges, it's very straightforward for users.

Reply:
I obviously like my `get_option(name, version : ...)` the most. I agree with Eli that there are some real gotchas to the version based approach. The keyword argument, while not perfect, gets the best of both approaches: It's explicit, it has 100% backwards compatibility, it's easy to implement, and it's easy to use, we don't end up with a proliferation of options. It would probably be a pretty easy feature for a linter to look for too.

The problem I have with renaming is it's simply not scalable. Pretty soon we'll end up with dozens of renamed options, sometimes with clunky names (`b_pie_feature`?). An end user then has to figure out that while they run `meson setup -Db_pie=disabled` that the have to in their script call `get_option('b_pie_feature')`. And then when we get to Meson 2.0, we have to decide what to do there. Do we pull a python and break everyone's build scripts in the short term, but have better long term maintainability (dict.iteritems() -> dict.items()), or do we remove the obvious name and get stuck with `get_option('b_pie_feature')` forever?

Reply:
> And not all projects use them at all!

Should a minimum meson version produce a warning of its own? And then no version is equivalent to '>=1.3'.

> It's explicit, it has 100% backwards compatibility, it's easy to implement, and it's easy to use,

I agree with everything except easy to use: the user still has to figure out that they write `-Db_pie=disabled` but they get false (and they can't distinguish auto from disabled) without adding the rare "version" keyword. It's not *totally* different from renaming in this respect, it's a better version of renaming but it shares some of the disadvantages.

Reply:
> Should a minimum meson version produce a warning of its own? And then no version is equivalent to '>=1.3'.

I'm very confused how this is supposed to solve the problem.

All projects without minimum versions that try to use this option would then ***immediately*** crash, instead of only crashing at some randomized date in the future. The goal isn't to have a flag day where everyone has to update their meson.build, but have that flag day be more predictable... the goal is to not have a flag day where lots of projects break.

Reply:
No, that would be if the default was `>=1.4`, not 1.3.